File: src/main/java/example/domain/model/specification/BasicTicket.java
package example.domain.model.specification;

import example.domain.model.bill.Amount;

/**
 * 乗車券
 */
public class BasicTicket {

    BasicFare basicFare;
    AdultType adultType;
    DepartureDate date;
    
    public BasicTicket(BasicFare basicFare, AdultType adultType, DepartureDate date) {
        this.basicFare = basicFare;
        this.adultType = adultType;
        this.date = date;
    }

    public Amount calculate() {
        return this.adultType.calculate(this.basicFare.calculate());
    }

    @Override
    public String toString() {
        return this.basicFare.destination + "行き " + adultType.toString() + " 料金" + calculate().toString() + "";
    }

}

File: jr-pricing/Model/BasicTicket.vb
Imports Jr.Pricing.Fw
Imports Jr.Pricing.Model.Bill
Imports Jr.Pricing.Model.Discount
Imports Jr.Pricing.Model.Specification

Namespace Model
    ''' <summary>
    ''' 乗車券
    ''' </summary>
    ''' <remarks></remarks>
    Public Class BasicTicket : Inherits ValueObject : Implements ITicket
        Private ReadOnly basicFare As BasicFare
        Private ReadOnly adultType As AdultType
        Private ReadOnly [date] As DepartureDate
        Private ReadOnly discounts As TicketDiscounts

        Public Sub New(basicFare As BasicFare, adultType As AdultType, [date] As DepartureDate)
            Me.new(basicFare, adultType, [date], Nothing)
        End Sub
        Public Sub New(basicFare As BasicFare, adultType As AdultType, [date] As DepartureDate, discounts As TicketDiscounts)
            Me.basicFare = basicFare
            Me.adultType = adultType
            Me.date = [date]
            Me.discounts = If(discounts, New TicketDiscounts)
        End Sub

        Public Function CalculateFare() As Amount Implements ITicket.CalculateFare
            Return discounts.Calculate(Me.adultType.CalculateFare(Me.basicFare.Calculate()))
        End Function

        ''' <summary>
        ''' 割引ルールを設定する
        ''' </summary>
        ''' <param name="discounts">割引ルール[]</param>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Function SetDiscounts(discounts As TicketDiscounts) As ITicket Implements ITicket.SetDiscounts
            Return New BasicTicket(basicFare, adultType, [date], discounts)
        End Function

        Protected Overrides Function GetAtomicValues() As IEnumerable(Of Object)
            Return New Object() {basicFare, adultType, [date]}
        End Function
    End Class
End Namespace